### DISCUSSION TITLE
Materialized view order of operations

### DISCUSSION THREAD

**Discussion by ckevin** [ClickHouse/ClickHouse#66723] [General]
If I have multiple materialized views triggered by the same source table, is there a way to control the order in which they execute?

I'm thinking of a situation in which I want to generate artificial UUIDv7 keys for a `ReplacingMergeTree` table (call it `D` for data) whose primary keys are poor for sort ordering (i.e. UUIDv4 for example.)  I want there to be a strict 1:1 relationship between the original primary key of the data, and a generated UUID -- a new value should only be generated when there has not been a UUIDv7 generated yet for that specific PK.

One thought I had was a MV to populate a `ReplacingMergeTree` of PK to UUID, (call it `M` for map) that would keep only the oldest version.  But I also want to trigger other MV's on inserts to `D`, and those MV's would want to be guaranteed that every row in `D` also has a row in the mapping table `M`.  If I could control the order in which the MV's run, to prioritize `M` first, that would be one way to solve it.  (I can't chain downstream MV's from `M`, since it won't have any other columns of interest.)

If I can't control MV ordering, I was also considering an approach where instead of having a separate mapping table `M`, the mapping is stored directly on `D` instead.  In that case, the first MV (for inserting data to `D`) uses an `IN` subquery join to look up existing data from `D` and either pulls in the existing UUID for each PK, or generates a new one if it is null, prior to inserting to `D`.  One nice thing about this approach is that I could potentially use that same self-join to emit +1/-1 versioned rows, and implement `VersionedCollapsingMergeTree` for compatibility with downstream aggregates.

Back to my original question though -- is it possible to rely on execution order of separate MV's off the same source table, or is it better to implement other patterns?

**Comment by den-crane** [comment]
They are executed in alphabetical order, if parallel execution is disabled.
But the problem that there is no guarantee that MV_B is able to read from a table produced by MV_A. Rows produced by the first MV may became visible only after the insert has finished.

**Reply by ckevin** [reply]
Thanks!  That makes sense about the row visibility too, I hadn't thought of that.

For my specific use case above, I realized that I probably have to use the self-join technique, because the UUIDv7 key should be used for table ordering on `D` (for better locality of data) and then a projection would allow lookups by the original (bad) key when needed.  I can't do that ordering if the generated keys are on a separate mapping table.

**Reply by JVMartin** [reply]
It may have been alphabetical at some point in the past, but it is not anymore, now it's by UUID.

Now, every MV gets a random UUID, which you can check with:
<img width="843" height="311" alt="image" src="https://github.com/user-attachments/assets/a87b61ac-8a6e-4ae9-8b92-42db55aa2bfb" />

Believe it or not, you have to keep deleting and re-creating the MVs until your dice rolls come up lucky enough for your MVs to be in the order you want. Absolutely ridiculous
